### PR TITLE
Add options to customize "MatchOutsideMap" behavior

### PR DESCRIPTION
**Pull request by JoaoBaptMG**
I decided to edit Tiled's source code after viewing [this discussion about the behavior of automapping near the map's edges](https://discourse.mapeditor.org/t/automapper-ignore-the-empty-void-beyond-the-perimeter-of-map/) and I was not satified with the state of the art of automapping near the borders.

Normally, when you try to automap near the edges, the edges are seen as empty tiles, failing to apply, for example, Remex's auto-generated rules (the ones originated from RPG Maker):
image

This pull request add two options (as map properties) to customize the behavior of automapping near the edges.
- **OverflowBorder**: this option makes the automapper treats the out-of-bounds areas of the map as if they were prolongations of the border tiles, thus applying the rules along the borders as if they used the same tiles as the nearest inbound tile. This is actually the default behavior of every RPG Maker since XP.
image
- **WrapBorder**: this option makes the automapper wrap the map around the edges, effectively applying the rules also to the other side of the border; this can be useful to create looped maps, to make it seamless around the borders:
image

I coded those two features so that if both **OverflowBorder** and **WrapBorder** are true, **WrapBorder** takes precedence. And, of course, it will not work with infinite maps, since the notion of "border" in them has no sense.

### CODE REVIEW

File: src/tiled/automapper.cpp [hunk 1]
```diff
@@ -631,8 +642,26 @@ static bool layerMatchesConditions(const TileLayer &setLayer,
...
+                        int height = setLayer.size().height();
+                        xd = (xd % width + width) % width;
+                        yd = (yd % height + height) % height;
+                    } else if (options.overflowBorder) {
+                        if (xd < 0) xd = 0;
+                        else if (xd >= setLayer.size().width())
+                            xd = setLayer.size().width() - 1;
+                        if (yd < 0) yd = 0;
+                        else if (yd >= setLayer.size().height())
+                            yd = setLayer.size().height() - 1;
```
Comment (bjorn) [inline src/tiled/automapper.cpp]: Please shorten this to:
```c++
xd = qBound(0, xd, setLayer.width() - 1);
yd = qBound(0, yd, setLayer.height() - 1);
```

File: src/tiled/automapper.cpp [hunk 2]
```diff
@@ -631,8 +642,26 @@ static bool layerMatchesConditions(const TileLayer &setLayer,
...
 
-                const Cell &setCell = setLayer.cellAt(x + offset.x(),
-                                                      y + offset.y());
+                int xd = x + offset.x();
+                int yd = y + offset.y();
+
+                if (!isMapInfinite) {
+                    if (options.wrapBorder) {
+                        int width = setLayer.size().width();
+                        int height = setLayer.size().height();
```
Comment (bjorn) [inline src/tiled/automapper.cpp]: Please use `setLayer.width()` and `setLayer.height()`. :-)

File: src/tiled/automapper.cpp [hunk 3]
```diff
@@ -631,8 +642,26 @@ static bool layerMatchesConditions(const TileLayer &setLayer,
                         !setLayer.contains(x + offset.x(), y + offset.y()))
                     return false;
 
-                const Cell &setCell = setLayer.cellAt(x + offset.x(),
-                                                      y + offset.y());
+                int xd = x + offset.x();
+                int yd = y + offset.y();
```
Comment (bjorn) [inline src/tiled/automapper.cpp]: Now that we have these variables, maybe move them up so that you can also use them in the `contains` check.

File: src/tiled/automapper.cpp (hunk 1 above)
Comment (JoaoBaptMG) [reply src/tiled/automapper.cpp]: Thanks! I have never written a line of Qt code before, so I didn't know of that function. That could be very useful, indeed!

**Comment by JoaoBaptMG** [comment]
>     * Would it make sense to have the OverflowBorder and WrapBorder options imply MatchOutsideMap = true? It feels superfluous to require the user to specify both.

I will incorporate this into my code as such, but I think it might be useful if the user disables MatchOutsideMap but forgets to do so with OverflowBorder/WrapBorder (well, that doesn't make any sense, but I guess it's a possible scenario?)

>     * Maybe instead of passing in whether the map is infinite, we can make sure these options are never enabled for infinite maps.

That could also be possible, and will remove excessive parameter passing.

File: src/tiled/automapper.cpp [hunk 4]
```diff
@@ -124,6 +139,17 @@ bool AutoMapper::setupRuleMapProperties()
...
+    // OverflowBorder and WrapBorder make no sense for infinite maps
+    if (mMapWork->infinite()) {
+        mOptions.overflowBorder = false;
+        mOptions.wrapBorder = false;
+    }
+
+    // Each of the border options imply MatchOutsideMap
+    if (mOptions.overflowBorder || mOptions.wrapBorder)
+        mOptions.matchOutsideMap = true;
+
```
Comment (JoaoBaptMG) [inline src/tiled/automapper.cpp]: As requested, I put the options to make the border options imply **MatchOutsideMap** and prevent infinite maps from turnin on the border options.

File: src/tiled/automapper.cpp (hunk 3 above)
Comment (JoaoBaptMG) [reply src/tiled/automapper.cpp]: Moved up.

File: src/tiled/automapper.cpp [hunk 5]
```diff
@@ -627,12 +653,24 @@ static bool layerMatchesConditions(const TileLayer &setLayer,
...
+                // so no "invalid" width/height accessing here.
+                if (options.wrapBorder) {
+                    xd = wrap(xd, setLayer.width());
+                    yd = wrap(yd, setLayer.height());
+                } else if (options.overflowBorder) {
+                    xd = qBound(0, xd, setLayer.width() - 1);
+                    yd = qBound(0, yd, setLayer.height() - 1);
+                }
+
+                const Cell &setCell = setLayer.cellAt(xd, yd);
```
Comment (JoaoBaptMG) [inline src/tiled/automapper.cpp]: Incorporated `qBound` as requested and add a helper function `wrap` to eliminate excessive code duplication.

File: src/tiled/automapper.cpp [hunk 6]
```diff
@@ -42,6 +42,11 @@
 
 using namespace Tiled;
 
+// Should i leave this function here?
+inline int wrap(int value, int bound) {
+    return (value % bound + bound) % bound;
+}
+
```
Comment (JoaoBaptMG) [inline src/tiled/automapper.cpp]: I haven't found a better place to this function, so I will leave it here.

File: src/tiled/automapper.cpp [hunk 7]
```diff
@@ -627,12 +653,24 @@ static bool layerMatchesConditions(const TileLayer &setLayer,
...
-                const Cell &setCell = setLayer.cellAt(x + offset.x(),
-                                                      y + offset.y());
+                // Those two options are guaranteed to be false if the map is infinite,
+                // so no "invalid" width/height accessing here.
+                if (options.wrapBorder) {
+                    xd = wrap(xd, setLayer.width());
+                    yd = wrap(yd, setLayer.height());
+                } else if (options.overflowBorder) {
+                    xd = qBound(0, xd, setLayer.width() - 1);
+                    yd = qBound(0, yd, setLayer.height() - 1);
```
Comment (JoaoBaptMG) [inline src/tiled/automapper.cpp]: Incorporated `qBound` as requested and add a helper function wrap to eliminate excessive code duplication.

File: src/tiled/automapper.cpp [hunk 8]
```diff
@@ -42,6 +42,11 @@
 
 using namespace Tiled;
 
+// Should i leave this function here?
+inline int wrap(int value, int bound) {
```
Comment (bjorn) [inline src/tiled/automapper.cpp]: The location is fine for now, but please define the function `static` and put the `{` on the next line.

Comment (JoaoBaptMG) [reply src/tiled/automapper.cpp]: `inline static` or `static inline`? I know there's no difference, but maybe there is a convention in Tiled's code?

File: src/tiled/automapper.cpp [hunk 9]
```diff
@@ -42,6 +42,12 @@
 
 using namespace Tiled;
 
+// Should i leave this function here?
+inline static int wrap(int value, int bound)
+{
```
Comment (JoaoBaptMG) [inline src/tiled/automapper.cpp]: Change done!

File: src/tiled/automapper.cpp (hunk 8 above)
Comment (bjorn) [reply src/tiled/automapper.cpp]: Actually I doubt there is any precedence here since I don't really see the point of defining a function both `inline` and `static`.

I actually only use `inline` when I need to use it. That is, to avoid multiple-definition problems for function definitions in header files that are outside class definitions. In any other case it's at best a hint that the compiler is free to ignore (and probably will, it'll inline only when it thinks it's a good idea anyway).

The `static` however is appropriate here, since we don't intend the function to be used from another translation unit. Defining it as static makes it inaccessible outside of its translation unit.

Comment (bjorn) [reply src/tiled/automapper.cpp]: > Actually I doubt there is any precedence here since I don't really see the point of defining a function both `inline` and `static`.

Maybe I should search the code before raising doubts. So, there's precedence for both `inline static` and `static inline`, with the latter a clear winner. However, of 9 total occurrences, 8 were inherited from 3rd party code. The one remaining one, I'm probably going to change to just `static` soon. :P

Comment (JoaoBaptMG) [reply src/tiled/automapper.cpp]: Not a problem, I will leave it just as `static`. Commiting the code again.

File: src/tiled/automapper.cpp (hunk 9 above)
Comment (JoaoBaptMG) [reply src/tiled/automapper.cpp]: Now it's just `static`

**Comment by bjorn** [comment]
@JoaoBaptMG Thanks again for this great improvement!

Would you like to have a go at updating [the documentation](https://github.com/bjorn/tiled/blob/master/docs/manual/automapping.rst#map-properties) for these attributes as well?

**Comment by JoaoBaptMG** [comment]
Yeah, I can do that. However, since the PR is already merged, should I open another PR for it or can I commit it to my fork?

**Comment by bjorn** [comment]
@JoaoBaptMG Yes, just open a new PR! :-)

**Comment by the-simian** [comment]
I'm late to the party here, but let me say this is awesome as can be, and I am so happy to see this work @JoaoBaptMG , @bjorn 

So very cool! WhenI work on my project create-phaser-app I will include this change and make that how the levels work
